File: lib/viewer.ml
open Async
open Misc

let signature debian =
  let open Deferred.Or_error.Let_syntax in
  let%bind () = check_command_exists "debsig-verify" in
  let%bind () = check_file_exists debian in

  let%bind.Deferred process =
    Process.create_exn ~prog:"debsig-verify"
      ~args:[ "--policies-dir"; "fake"; debian ]
      ()
  in
  let%bind.Deferred output = Process.collect_output_and_wait process in
  match output.exit_status with
  | Ok _ ->
      let msg =
        Printf.sprintf
          "Cannot look up package signature due to internal error. Expecting \
           command to error out \n\
          \ %s" output.stdout
      in
      Log.Global.error "%s \n" msg ;
      Deferred.Or_error.error_string msg
  | Error _ -> (
      let pattern = Str.regexp ".*fake/\\([A-Z0-9]+\\):.*" in
      match Str.string_match pattern output.stdout 0 with
      | true ->
          let extracted_id = Str.matched_group 1 output.stdout in
          Deferred.Or_error.return extracted_id
      | false ->
          Deferred.Or_error.error_string "Failed to extract ID from output" )

File: src/viewer.rs
use anyhow::{anyhow, Result};
use regex::Regex;
use std::process::Command;

use crate::misc::{check_command_exists, check_file_exists};

/// Extract the signing-key id from a .deb by invoking debsig-verify with a
/// deliberately-fake policies directory and parsing the resulting error.
///
/// debsig-verify prints lines like:
///   debsig: Origin Signature check failed.  This deb might not be signed,
///   ...
///   fake/<KEY_ID>: ...
/// — the path `fake/<KEY_ID>` is what we grep for.
pub fn signature(deb: &str, debug: bool) -> Result<String> {
    check_command_exists("debsig-verify")?;
    check_file_exists(deb)?;

    if debug {
        log::info!("Executing: debsig-verify --policies-dir fake {}", deb);
    }

    let output = Command::new("debsig-verify")
        .args(["--policies-dir", "fake", deb])
        .output()
        .map_err(|e| anyhow!("Failed to spawn debsig-verify: {}", e))?;

    if output.status.success() {
        let stdout = String::from_utf8_lossy(&output.stdout);
        let msg = format!(
            "Cannot look up package signature due to internal error. Expecting \
             command to error out\n {}",
            stdout
        );
        log::error!("{}", msg);
        return Err(anyhow!(msg));
    }

    // Newer versions of debsig-verify route the diagnostic containing
    // the `fake/<KEY_ID>` path to stderr; older versions printed it to
    // stdout. Search both streams so the parse stays working across
    // distro versions.
    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);
    let re = Regex::new(r"fake/([A-Fa-f0-9]+)[:/]").unwrap();
    if let Some(caps) = re.captures(&stdout).or_else(|| re.captures(&stderr)) {
        return Ok(caps.get(1).unwrap().as_str().to_uppercase());
    }
    Err(anyhow!(
        "Failed to extract key id from debsig-verify output.\n\
         stdout: {}\n\
         stderr: {}",
        stdout.trim(),
        stderr.trim()
    ))
}
